dom5.js

```diff
@@ -455,9 +457,63 @@ function newElement(tagName, namespace) {
   };
 }
 
+function newDocumentFragment() {
+  return {
+    nodeName: '#document-fragment',
+    childNodes: [],
+    parentNode: null,
+  };
+}
+
+function cloneNode(node) {
+  // parent is a backreference, and we don't want to clone the whole tree, so
+  // make it null before cloning.
+  var parent = node.parentNode;
+  node.parentNode = null;
+  var clone = cloneObject(node);
+  node.parentNode = parent;
+  return clone;
+}
+
+/**
+ * Inserts `newNode` into `parent` at `index`, optionally replaceing the
+ * current node at `index`. If `newNode` is a DocumentFragment, its childNodes
+ * are inserted and removed from the fragment.
+ */
+function insertNode(parent, index, newNode, replace) {
```

Comment:
I added this so that common insertion logic, like handling fragments and removing new nodes from their old parents, is handled in one place. I didn't add any tests for this function directly, but it's tested via replace, append and insertBefore.